doc/vxlan/Overlay ECMP with BFD.md

```diff
@@ -0,0 +1,216 @@
+|                          |                                |
+|--------------------------|--------------------------------|
+| BFD                      | Bidirectional Forwarding       |
+| VNI                      | Vxlan Network Identifier       |
+| VTEP                     | Vxlan Tunnel End Point         |
+| VNet                     | Virtual Network                |
+
+# 1 Requirements Overview
+
+## 1.1 Usecase
+
+![](https://github.com/Azure/SONiC/blob/master/images/vxlan_hld/OverlayEcmp_UseCase.png)
+
+## 1.1 Functional requirements
+
+At a high level the following should be supported:
+
+- Configure ECMP with Tunnel Nexthops (IPv4 and IPv6)
+- Tunnel Endpoint monitoring via BFD
+- Add/Withdraw Nexthop based on Tunnel or Endpoint health
+
+## 1.2 CLI requirements
+- User should be able to show the BFD session
+- User should be able to show the Vnet routes
+
+## 1.3 Warm Restart requirements
+No special handling for Warm restart support.
+
+# 2 Modules Design
+
+The following are the schema changes. 
+
+## 2.1 Config DB
+
+Existing Vxlan and Vnet tables. 
+
+### 2.1.1 VXLAN Table
+```
+VXLAN_TUNNEL|{{tunnel_name}} 
+    "src_ip": {{ip_address}} 
+    "dst_ip": {{ip_address}} (OPTIONAL)
+```
+### 2.1.2 VNET/Interface Table
+```
+VNET|{{vnet_name}} 
+    "vxlan_tunnel": {{tunnel_name}}
+    "vni": {{vni}} 
+    "scope": {{"default"}} (OPTIONAL)
+    "peer_list": {{vnet_name_list}} (OPTIONAL)
+```
+
+## 2.2 APP DB
+
+### VNET
+
+The following are the changes for Vnet Route table
+
+Existing:
+
+``` 
+VNET_ROUTE_TUNNEL_TABLE:{{vnet_name}}:{{prefix}} 
+    "endpoint": {{ip_address}} 
+    "mac_address":{{mac_address}} (OPTIONAL) 
+    "vni": {{vni}}(OPTIONAL) 
+```
+
+Proposed:
+```
+VNET_ROUTE_TUNNEL_TABLE:{{vnet_name}}:{{prefix}}  
+    "endpoint": {{ip_address1},{ip_address2},...} 
+    "endpoint_monitor": {{ip_address1},{ip_address2},...} (OPTIONAL) 
+    "mac_address":{{mac_address1},{mac_address2},...} (OPTIONAL)  
+    "vni": {{vni1},{vni2},...} (OPTIONAL) 
+    "weight": {{w1},{w2},...} (OPTIONAL) 
+    “profile”: {{profile_name}} (OPTIONAL) 
+```
+
+```
+key                      = VNET_ROUTE_TUNNEL_TABLE:vnet_name:prefix ; Vnet route tunnel table with prefix 
+; field                  = value 
+ENDPOINT                 = list of ipv4 addresses    ; comma separated list of endpoints 
+ENDPOINT_MONITOR         = list of ipv4 addresses    ; comma separated list of endpoints 
+MAC_ADDRESS              = 12HEXDIG                  ; Inner dst mac in encapsulated packet 
+VNI                      = DIGITS                    ; VNI value in encapsulated packet 
+WEIGHT                   = DIGITS                    ; Weights for the nexthops, comma separated (Optional) 
+PROFILE                  = STRING                    ; profile name to be applied for this route, for community  
+                                                       string etc (Optional) 
+```
+
+### BFD
+
+```
+BFD_SESSION:{{ifname}}:{{prefix}}  
```

Comment:
In control plane(FRR) BFD, session configuration is not directly associated to interface(interface is an optional key param), and a multihop session cannot be directly associated BFD with the interface. If we want to harmonize with control plane BFD we should consider having BFD session key same as control plane.

Reply:
We thought about it and for APP_DB and BFD orch, we think having ifname and ipaddr would be straight-forward. For the config_db based session creation, it can be session_name based (as key), and BFD manager can create the APP_DB sessions and resolve if there are multiple sessions with same destination etc.